infra/Pos/Slotting/MemState/Holder.hs

```diff
@@ -42,22 +50,80 @@ class HasSlottingVar ctx where
 type SlotsDefaultEnv ctx m =
     (MonadReader ctx m, HasSlottingVar ctx, MonadIO m)
 
-getSystemStartDefault :: SlotsDefaultEnv ctx m => m Timestamp
-getSystemStartDefault = view slottingTimestamp
+-- A common function we extracted. It executes a pure function that requires
+-- @SlottingData@ and returns the result in @STM@ monad.
+withSlottingVar
+    :: SlotsDefaultEnv ctx m
+    => (SlottingData -> b)
+    -> m b
+withSlottingVar f = do
+    var <- view slottingVar
+    atomically $ f <$> readTVar var
+
+getSystemStartDefault
+    :: SlotsDefaultEnv ctx m
+    => m Timestamp
+getSystemStartDefault =
+    view slottingTimestamp
+
+getAllEpochIndicesDefault
+    :: SlotsDefaultEnv ctx m
+    => m [EpochIndex]
+getAllEpochIndicesDefault =
+    withSlottingVar getAllEpochIndices
 
-getSlottingDataDefault :: SlotsDefaultEnv ctx m => m SlottingData
-getSlottingDataDefault = atomically . readTVar =<< view slottingVar
+-- | We get the current and the next epoch index atomically since we don't want to
+-- allow a chance that a timing issue messes up the indexes - to return say (13,13).
+getCurrentNextEpochIndexDefault
+    :: SlotsDefaultEnv ctx m
+    => m (EpochIndex, EpochIndex)
+getCurrentNextEpochIndexDefault =
+    withSlottingVar getCurrentNextIndex
+  where
+    getCurrentNextIndex :: SlottingData -> (EpochIndex, EpochIndex)
+    getCurrentNextIndex slottingData =
+        (getCurrentEpochIndex slottingData, getNextEpochIndex slottingData)
 
-waitPenultEpochEqualsDefault :: SlotsDefaultEnv ctx m => EpochIndex -> m ()
-waitPenultEpochEqualsDefault target = do
+-- | As with the function above, we don't want to get into a situation where we
+-- can have timing issues and recieve wrong results.
+getCurrentNextEpochSlottingDataDefault
+    :: SlotsDefaultEnv ctx m
+    => m (EpochSlottingData, EpochSlottingData)
+getCurrentNextEpochSlottingDataDefault =
+    withSlottingVar getCurrentNextEpochSlottingData
+  where
+    getCurrentNextEpochSlottingData
+        :: SlottingData
+        -> (EpochSlottingData, EpochSlottingData)
+    getCurrentNextEpochSlottingData sd =
+        (getCurrentEpochSlottingData sd, getNextEpochSlottingData sd)
+
+getEpochSlottingDataDefault
+    :: SlotsDefaultEnv ctx m
+    => EpochIndex
+    -> m (Maybe EpochSlottingData)
+getEpochSlottingDataDefault ei =
+    withSlottingVar $ lookupEpochSlottingData ei
+
+-- TODO(ks): This is also quite unsafe. We should switch to @addEpochSlottingDataDefault@.
```

Comment:
Yes, but then we need to change this too - https://github.com/input-output-hk/cardano-sl/pull/1217#discussion_r133762869
The thing I'm concerned with is the rollback and what will happen with the data then.